File: build-tools/builder-configuration-cache/src/main/java/org/gradle/configurationcache/VintageBuildTreeLifecycleControllerFactory.java
package org.gradle.configurationcache;

import org.gradle.internal.operations.BuildOperationExecutor;
import org.gradle.internal.resources.ProjectLeaseRegistry;
import org.gradle.composite.internal.BuildTreeWorkGraphController;
import org.gradle.internal.build.BuildLifecycleController;
import org.gradle.internal.buildtree.BuildModelParameters;
import org.gradle.internal.buildtree.BuildTreeFinishExecutor;
import org.gradle.internal.buildtree.BuildTreeLifecycleController;
import org.gradle.internal.buildtree.BuildTreeLifecycleControllerFactory;
import org.gradle.internal.buildtree.BuildTreeWorkExecutor;
import org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController;
import org.gradle.internal.buildtree.DefaultBuildTreeModelCreator;
import org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer;
import org.gradle.internal.model.StateTransitionControllerFactory;

public class VintageBuildTreeLifecycleControllerFactory implements BuildTreeLifecycleControllerFactory {

    private final BuildModelParameters buildModelParameters;
    private final BuildTreeWorkGraphController taskGraph;
    private final BuildOperationExecutor buildOperationExecutor;
    private final ProjectLeaseRegistry projectLeaseRegistry;
    private final StateTransitionControllerFactory stateTransitionControllerFactory;

    public VintageBuildTreeLifecycleControllerFactory(BuildModelParameters buildModelParameters,
                                                      BuildTreeWorkGraphController taskGraph,
                                                      BuildOperationExecutor buildOperationExecutor,
                                                      ProjectLeaseRegistry projectLeaseRegistry,
                                                      StateTransitionControllerFactory stateTransitionControllerFactory) {
        this.buildModelParameters = buildModelParameters;
        this.taskGraph = taskGraph;
        this.buildOperationExecutor = buildOperationExecutor;
        this.projectLeaseRegistry = projectLeaseRegistry;
        this.stateTransitionControllerFactory = stateTransitionControllerFactory;
    }

    @Override
    public BuildTreeLifecycleController createRootBuildController(BuildLifecycleController targetBuild,
                                                                  BuildTreeWorkExecutor workExecutor,
                                                                  BuildTreeFinishExecutor finishExecutor) {
        return createController(targetBuild, workExecutor, finishExecutor);

    }

    @Override
    public BuildTreeLifecycleController createController(BuildLifecycleController targetBuild,
                                                         BuildTreeWorkExecutor workExecutor,
                                                         BuildTreeFinishExecutor finishExecutor) {
        DefaultBuildTreeWorkPreparer workPreparer = createWorkPreparer(targetBuild);
        DefaultBuildTreeModelCreator modelCreator = createModelCreator(targetBuild);
        return new DefaultBuildTreeLifecycleController(targetBuild, taskGraph, workPreparer, workExecutor, modelCreator, finishExecutor, stateTransitionControllerFactory);
    }


    private DefaultBuildTreeModelCreator createModelCreator(BuildLifecycleController targetBuild) {
        return new DefaultBuildTreeModelCreator(buildModelParameters, targetBuild.getGradle().getOwner(), buildOperationExecutor, projectLeaseRegistry);
    }

    private DefaultBuildTreeWorkPreparer createWorkPreparer(BuildLifecycleController targetBuild) {
        return new DefaultBuildTreeWorkPreparer(targetBuild.getGradle().getOwner(), targetBuild);
}
}

File: build-tools/builder-configuration-cache/src/main/java/org/gradle/configurationcache/VintageBuildTreeLifecycleControllerFactory.kt
package org.gradle.configurationcache

import org.gradle.composite.internal.BuildTreeWorkGraphController
import org.gradle.internal.build.BuildLifecycleController
import org.gradle.internal.buildtree.BuildModelParameters
import org.gradle.internal.buildtree.BuildTreeFinishExecutor
import org.gradle.internal.buildtree.BuildTreeLifecycleController
import org.gradle.internal.buildtree.BuildTreeLifecycleControllerFactory
import org.gradle.internal.buildtree.BuildTreeWorkExecutor
import org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController
import org.gradle.internal.buildtree.DefaultBuildTreeModelCreator
import org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer
import org.gradle.internal.model.StateTransitionControllerFactory
import org.gradle.internal.operations.BuildOperationExecutor
import org.gradle.internal.resources.ProjectLeaseRegistry


class VintageBuildTreeLifecycleControllerFactory(
    private val buildModelParameters: BuildModelParameters,
    private val taskGraph: BuildTreeWorkGraphController,
    private val buildOperationExecutor: BuildOperationExecutor,
    private val projectLeaseRegistry: ProjectLeaseRegistry,
    private val stateTransitionControllerFactory: StateTransitionControllerFactory
) : BuildTreeLifecycleControllerFactory {
    override fun createRootBuildController(targetBuild: BuildLifecycleController, workExecutor: BuildTreeWorkExecutor, finishExecutor: BuildTreeFinishExecutor): BuildTreeLifecycleController {
        return createController(targetBuild, workExecutor, finishExecutor)
    }

    override fun createController(targetBuild: BuildLifecycleController, workExecutor: BuildTreeWorkExecutor, finishExecutor: BuildTreeFinishExecutor): BuildTreeLifecycleController {
        val workPreparer = createWorkPreparer(targetBuild)
        val modelCreator = createModelCreator(targetBuild)
        return DefaultBuildTreeLifecycleController(targetBuild, taskGraph, workPreparer, workExecutor, modelCreator, finishExecutor, stateTransitionControllerFactory)
    }

    fun createModelCreator(targetBuild: BuildLifecycleController) =
        DefaultBuildTreeModelCreator(buildModelParameters, targetBuild.gradle.owner, buildOperationExecutor, projectLeaseRegistry)

    fun createWorkPreparer(targetBuild: BuildLifecycleController) = DefaultBuildTreeWorkPreparer(targetBuild.gradle.owner, targetBuild)
}
